Add systemtap probes into the top-level data provider requests

justin-stephenson opened the pull request:
Adding Data Provider probes and SystemTap script **dp_request.stp** to probe 
DP requests for performance analysis primarily based on elapsed time to complete a request.
The intention is to identify delays during logins and user lookups.

I'm not sure if the comments in **src/systemtap/sssd.stp.in** are worthwhile but the intention for someone writing their own stap script they could avoid having to look at the actual source code to see which target variables are retrieved(though the variable names are self-explanatory).

Resolves: https://pagure.io/SSSD/sssd/issue/3061

justin-stephenson commented:
I also wrote a short blog post about using this SystemTap script

http://justin-stephenson.blogspot.com/2017/05/measuring-sssd-performance-with.html

sumit-bose commented:
ok to test

jhrozek commented:
So far I've only scrolled through the patches and they look OK conceptually, but a bit more testing is needed.

However, could you please add your blog to planet freeipa? http://planet.freeipa.org/

I think Petr Vobornik from the IPA team could add you to the aggregator.

justin-stephenson commented:
Thank you for the review Jakub.

I only added 2 Probe markers in the code to get the high-level request information, I can add them into more places for this PR if you would like.

jhrozek commented:
On Mon, May 08, 2017 at 07:03:16AM -0700, Justin Stephenson wrote:
> Thank you for the review Jakub.
> 
> I only added 2 Probe markers in the code to get the high-level request information, I can add them into more places for this PR if you would like.
> 

The question I'm thinking about is -- do we want to make the probes
exposed also for admins or support engineers?

If yes, then I think it would be nice if the probes were documented and at
least a bit stable. Of course not stable like we do keep our library APIs,
but more stable than our DEBUG messages.

I'm leaning towards documenting some of the probes. The probes at points
that correspond to fundamental sssd architectural decisions (data
provider, responding to a request from cache, ...) could IMO be
documented and we could even have a sssd-systemtap man page.

The probes that are in parts of code that are subject to change (like
the current probes that we have in the nested LDAP group resolution
code) don't have to be documented. Then the only way for an admin to
find about the probes is to check the code or the files under
/usr/share/systemtap -- and if the admin is able to do that, they are
able to cope with changes between versions.

I'm really interested to hear what do other developers think, though.

justin-stephenson commented:
@jhrozek thanks a lot for your input and sorry for the delayed response. I wanted to give a chance for other developers to provide any comments and I know everyone is busy.

I will work on documenting the high-level probes as you mention in a `sssd-systemtap` or `sssd-stap` man page and update the PR accordingly. Based on this, does it make sense to add tests to ensure stability ?

If you have other thoughts/suggestions i'm open to those, please go ahead and add **Changes Requested** (as it looks like I cannot)

justin-stephenson commented:
@jhrozek I made the changes requested and added a **sssd-systemtap** man page.

Regarding the man page - I am not sure if the description of each probe point should be more verbose, or if the exact definition should be added for the `probestr` variable.

I created a new ticket for the more handler/provider specific systemtap code probe additions.

https://pagure.io/SSSD/sssd/issue/3416

lslebodn commented on src/systemtap/sssd_functions.stp:
```diff
@@ -64,3 +64,61 @@ function sssd_acct_req_probestr(fc_name, entry_type, filter_type,
+    } else if (target == 8) {
+        str_target = "SUBDOMAINS"
+    } else if (target == 9) {
+        str_target = "TARGET_SENTINEL"
```
I would prefer to define "constants" in the beginning of file.

`/usr/share/systemtap/examples/process/schedtimes.stp` has something like this:
```
//constants
global DEAD=-1, RUNNING=1, QUEUED=2, SLEEPING=3
```

I am not sure whether these constant will be reused but one never know :-)

justin-stephenson commented:
@lslebodn added constant definitions as suggested, thank you.

fidencio commented:
Removing the "Changes Requested" label as Justin updated the patch set a few weeks ago.

jhrozek commented:
I'm really sorry this review took so long. I rebased the patches (I'll push them to this PR) and they look good to me. Here is example output:
```
stap /usr/share/sssd/systemtap/dp_request.stp                                                      
        *** Beginning run! ***                             
        --> DP Request [Account #5] sent for domain [win.trust.test]                                                  
        --> Target: [ID] - Method: [Account Handler]       
                 DP Request [Account #5] finished with return code [0]: [Success]                                     
                 Elapsed time [0m0.003s]                   

        --> DP Request [Account #6] sent for domain [win.trust.test]                                                  
        --> Target: [ID] - Method: [Account Handler]       
                 DP Request [Account #6] finished with return code [0]: [Success]                                     
                 Elapsed time [0m0.002s]                   

        --> DP Request [Account #7] sent for domain [child.win.trust.test]                                            
        --> Target: [ID] - Method: [Account Handler]       
                 DP Request [Account #7] finished with return code [0]: [Success]                                     
                 Elapsed time [0m0.091s]                   

        --> DP Request [Account #8] sent for domain [sibling.win.trust.test]                                          
        --> Target: [ID] - Method: [Account Handler]       
                 DP Request [Account #8] finished with return code [0]: [Success]                                     
                 Elapsed time [0m0.047s]                   

^C
Ending Systemtap Run - Providing Summary
Total Number of DP requests: [4]
Total time in DP requests: [0m0.143s]
Slowest request data:
        Request: [Account #7]
        Target:  [ID]
        Method:  [Account Handler]
        Start Time: [Tue Sep  5 13:01:56 2017 UTC]
        End Time: [Tue Sep  5 13:01:56 2017 UTC]
        Duration: [0m0.091s]
```

lslebodn commented:
Is there any reason why `%{_mandir}/man5/sssd-systemtap.5*` is added "twice" into sssd-common?
and I assume we do not want to install that man page with `--disable-systemtap`.

BTW I cannot see referenced this page  in "SEE ALSO" section. Do we want to add it there?

justin-stephenson commented:
Rebased this and addressed comments from @lslebodn 

With these changes I tested `reconfig --disable-systemtap` and `reconfig --enable-systemtap=no` but I am still seeing the sssd-systemtap man page, I would appreciate any help on this if I am missing something here in the Makefile changes.

lslebodn commented:
@justin-stephenson. I am not sure what did you mean by last comment. Because last version works for me.

```
sh$ ./configure 
sh$ make V=0 -j8
sh$ make install V=0 DESTDIR=$PWD/_inst
sh$ find _inst/ -name "sssd-systemtap*"
```
vs

```
sh$ ./configure --enable-systemtap
sh$ make V=0 -j8
sh$ make install V=0 DESTDIR=$PWD/_inst
sh$ find _inst/ -name "sssd-systemtap*"
_inst/usr/local/share/man/man5/sssd-systemtap.5
```

Could you elaborate a little bit? Maybe I overlooked something.

justin-stephenson commented:
@lslebodn it was a mistake in my local testing I am sure. If it works for you then please consider this set of patches ready for review and approval.

lslebodn commented:
master:
* f199c749197532fd3380fa6f5d9f7a579879c21a
* 1182dd93a5a6bb18943284273f7fd59b83468843
* d46d59e78600aa72176df7217c94743b7e71881a

sssd-1-14:
* 4cf74be9a2fd3c3e56feeac719426f9b08a0bad2
* 6e579bf857faa9e9b02da355146ccccf70bd77c6
* 5df609bb8eeee157892acf9f622b2c0143dc9010
